core/src/actors/chain_manager/handlers.rs

```diff
@@ -71,37 +72,59 @@ impl Handler<GetHighestCheckpointBeacon> for ChainManager {
 
 /// Handler for AddNewBlock message
 impl Handler<AddNewBlock> for ChainManager {
-    type Result = Result<Hash, ChainManagerError>;
-
-    fn handle(
-        &mut self,
-        msg: AddNewBlock,
-        _ctx: &mut Context<Self>,
-    ) -> Result<Hash, ChainManagerError> {
-        let res = self.process_new_block(msg.block);
-        match res {
-            Ok(hash) => {
-                // Get SessionsManager's address
-                let sessions_manager_addr = System::current().registry().get::<SessionsManager>();
-
-                // Tell SessionsManager to announce the new block through every consolidated Session
-                let items = vec![InventoryEntry::Block(hash)];
-                sessions_manager_addr.do_send(Broadcast {
-                    command: AnnounceItems { items },
-                });
-            }
-            Err(ChainManagerError::BlockAlreadyExists) => {
-                debug!("Block already exists");
-            }
-            Err(ChainManagerError::StorageError(_)) => {
-                debug!("Error when serializing block");
-            }
-            Err(_) => {
-                debug!("Unexpected error");
-            }
-        };
-
-        res
+    type Result = SessionUnitResult;
+
+    fn handle(&mut self, msg: AddNewBlock, ctx: &mut Context<Self>) {
+        // Block verify process
+        let reputation_manager_addr = System::current().registry().get::<ReputationManager>();
+
+        let block_epoch = msg.block.header.block_header.beacon.checkpoint;
+        if self.current_epoch == None {
+            warn!("ChainManager doesn't have current epoch");
+        } else if Some(block_epoch) != self.current_epoch {
+            warn!("Block epoch not valid");
+        } else if !self.validate_coinbase() {
+            warn!("Block coinbase not valid");
+        } else if !self.validate_merkle_tree() {
+            warn!("Block merkle tree not valid");
+        } else {
+            // Request proof of eligibility validation to ReputationManager
+            reputation_manager_addr
+                .send(ValidatePoE {
+                    block_header: msg.block.header.clone(),
+                })
+                .into_actor(self)
+                .then(|res, act, _ctx| {
+                    match res {
+                        Err(e) => {
+                            // Error when sending message
+                            error!("Unsuccessful communication with reputation manager: {}", e);
+                        }
+                        Ok(valid_poe) => {
+                            if !valid_poe {
```

Comment:
Even if checking for errors first looks like a good idea from a _fail-fast_ mindset, I think that checking for positive assertions should be always preferred for the sake of readability:
```suggestion
                            if valid_poe {
```
Otherwise, the triggering logic for the `else` clause could become not evident, because it's a double negation.

Also, because of optimization, it's not a bad idea to expect the first branch in an assertion to be the most common one. In addition, we save the `!` operator, which comes at a mostly negligible yet different-from-zero cost.

Only case when `if !expression` is acceptable IMO is when we have no `else` clause.

What do you think? @witnet/devs

Reply:
When checking for multiple failure conditions, it makes sense to leave the success branch as the last one.

I like `if !valid { return err; }`, as it eliminates the need for an else branch, but in this case it can be nicely replaced with:

```rust
                    match res {
                        Err(e) => {
                            // Error when sending message
                            error!("Unsuccessful communication with reputation manager: {}", e);
                        }
                        Ok(false) => {
                            warn!("Block PoE not valid"); 
                        }
                        Ok(true) => {
                            // Valid
                        }
```

Reply:
Nice!!